text/0000-grid-identity.md

```diff
@@ -0,0 +1,718 @@
+t-shirt tank management company. After the appropriate retrimands are dealt,
+the Beta Company deactivates their previous `beta.Drivers` role and creates
+two new roles, one for each tank company.
+
+```yaml
+- name: beta.Drivers
+  permissions:
+    - tankops::can-drive
+    - tankops::can-turn-turret
+    - tankops::can-fire
+    - tankops::can-decommission
+  allowed_organizations: []
+  inherit_from:
+    - alpha.Drivers
+    - delta.TankOperator
+  active: false
+```
+
+```yaml
+- name: beta.AlphaDrivers
+  permissions:
+    - tankops::can-drive
+    - tankops::can-turn-turret
+    - tankops::can-fire
+  allowed_organizations: []
+  inherit_from:
+    - alpha.Drivers
+  active: true
+```
+
+```yaml
+- name: beta.DeltaDrivers
+  permissions:
+    - tankops::can-drive
+    - tankops::can-turn-turret
+    - tankops::can-fire
+    - tankops::can-decommission
+  allowed_organizations: []
+  inherit_from:
+    - delta.TankOperator
+  active: true
+```
+
+After this embarassing incident, the t-shirt tank consortium sets to work on a
+new version of the tank operation smart contract to better handle exclusivity
+agreements.
+
+## Alternate IDs
+
+In several Grid smart contracts, it is necessary to record an additional ID to
+identify an organization in a certain context. For example, in Grid Product, the
+organization must have the proper GS1 company prefix in order to interact with
+products with correlating GTINs. In Grid Pike, this is handled through the use
+of the metadata field.
+
+In Grid Identity, we further define this functionality through a new concept
+called "Alternate IDs". This is implemented through a field called
+`alternate_ids` on the `Organization` message and a new message called
+`AlternateIDIndexEntry`. The `alternate_ids` field allows users and smart
+contracts to easily fetch a list of all of the alternate IDs associated with an
+organization. The AlternateIDIndexEntry message serves as an index to fetch a
+Grid Identity organization from an externally known ID, and also ensures that no
+organization can have an alternate ID that is already assigned.
+
+When an organization is created, we check the list of alternate IDs supplied in
+the transaction to ensure that none of the alternate IDs are already associated
+with an existing organization. If not, we create a new `AlternateIDIndexEntry`
+in state for each alternate ID. If that organization is updated, we check the
+list of alternate IDs supplied with the transaction against the present record.
+Any missing alternate IDs will be deleted from the index (the
+`AlternateIDIndexEntry` for that alternate ID is deleted) and any new IDs will
+be checked and added.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## State
+
+Agents, organizations, and roles are stored in state using protobuf message
+format. These messages are defined as follows:
+
+```protobuf
```

Comment:
regarding the `active` attribute. This is listed below in the `Agent` message but not in the `Role` message. The tank examples above use an `active` field in each `Role`, however.
In either the Agent or Role case, why would the field be inactive rather than removed from state?

Reply:
Great catch, I had intended to add the "active" field to the protobuf. However this has brought up some questions in my mind.

Regarding the necessity of the field, for Agents I believe it is necessary in any case where a smart contract might store the creator or custodian of some record in state. If an agent is no longer active (the person leaves the org or some such) we can preserve that relationship by setting the agent to inactive instead of deleting it. This makes less sense if we also have the option to delete though.  This argument is not as strong for roles though.